### ISSUE TITLE
Changes to spec.source.helm.values and spec.source.helm.parameters do not trigger Job update

### ISSUE THREAD

**Issue by gberenice** [argoproj/argo-cd#6767]
Checklist:

* [X] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [X] I've included steps to reproduce the bug.
* [X] I've pasted the output of `argocd version`.

**Describe the bug**

We use ArgoCD Application to manage our custom helm chart that includes k8s Job. We pass some values (like environment variables) to that job via `spec.source.helm.values` and `spec.source.helm.parameters`. When we push changes to git for `spec.source.helm.values` or `spec.source.helm.parameters` we see corresponding changes in Application manifest, but new Job doesn't start. Forcing sync by pressing Synchronize button helps - previous one is removed and new one with updated parameters is created.

Any other change (like `spec.source.helm.releaseName`) works as expected.

Job hooks:
```shell
    argocd.argoproj.io/hook: PreSync
    argocd.argoproj.io/hook-delete-policy: BeforeHookCreation
```

Application sync configuration:
```shell
      syncPolicy:
        automated:
          prune: true 
          selfHeal: true
        syncOptions:
        - CreateNamespace=false
        - ApplyOutOfSyncOnly=true
        - PruneLast=true
```
We've tried different syncOptions, including `ApplyOutOfSyncOnly=false`, but it doesn't change this behaviour.

**To Reproduce**

- Apply Application that manages helm chart with Job that expects some values to be passed
- Update/add any value in `spec.source.helm.parameters`
- See that Application live manifest is updated, but all is `Synced`.

**Expected behavior**

`OutOfSync` is detected, new Job with updated parameters is started.

**Version**

```shell
argocd: v2.0.3+8d2b13d
  BuildDate: 2021-05-27T17:38:37Z
  GitCommit: 8d2b13d733e1dff7d1ad2c110ed31be4804406e2
  GitTreeState: clean
  GoVersion: go1.16
  Compiler: gc
  Platform: linux/amd64
```

**Comment by charles-woshicai** [comment]
> Forcing sync by pressing Synchronize button helps - previous one is removed and new one with updated parameters is created.

This is strange, as I know argo-cd application controller would auto-sync application per 3min by default, and it should be equal to click `Sync` button.(please be free to correct me if there is any incorrect).

**Comment by gberenice** [comment]
@charles-woshicai yes, it should, and it works for the rest of `spec` updates. But for described particular `spec` change it doesn't.

**Comment by taintedkernel** [comment]
I've noticed this as well.  In my case, I was trying the legacy minio chart (at https://github.com/minio/charts) and on version 2.1.2.

I created the application and populated a values file, but did not sync.  This issue was noticeable as I was scanning the diff before syncing - no job appeared in the rendered diff.  A local `helm template` run did show the expected job, so I'm confident this is related to Argo.
